src/plot.js

```diff
@@ -96,12 +95,38 @@ export function plot(options = {}) {
       .call(applyInlineStyles, style)
     .node();
 
-  for (const mark of marks) {
+  // A mark layout is responsible for:
+  // * applying the scales, possibly pushing a new one (for legends)
+  // * returning a filtered index and values to render, possibly new channels to rescale
+  const newScaleChannels = new Map();
+  const layouts = [];
```

Comment:
This array overlaps with markIndex: layouts[i].index is equal to markIndex.get(marks[i]) except in the case where the layout returned a new index. We should ideally normalize the state so there are no duplicate (and obsolete or inconsistent) representations.